### ISSUE TITLE
Static libraries are impossibly big

### ISSUE THREAD

**Issue by Blacklands** [cginternals/glbinding#189]
Hi,
I'm currently experiencing a problem with building glBinding from the latest source.
Before I start, let me just say that I'm very new to building C++ libraries, so it's likely that I am just doing this wrong.
What I did was getting the newest source, running CMake to generate a Visual Studio 2015 project, and compiling that. I do this because I'd like to have the statically linked version of glBinding, and it seems that there is only the DLL version available pre-built (unless I missed something).
Now I know that static libraries tend to be a bit bigger than dynamic ones. But surely this:
![over 150 mbs in size](https://cloud.githubusercontent.com/assets/4943543/16561746/261b5daa-41fb-11e6-9b1d-d00c6118802d.PNG)
cannot possibly be correct?!
Especially considering that the DLL versions are around 5.5 and 15.5 MBs (Release and Debug respective) for me (when I build them), which seems to be a reasonable size to me.

Here's a more detailed rundown of what I am doing:

1. Clone glbinding-master from GitHub.
2. Fire up CMake. I normally use the GUI for it, but I couldn't get that to work with glBinding so I'm using the command line now. I have the newest version of it.
3. Create a "build" folder, type `cmake .. -G "Visual Studio 14 2015 Win64"`, then change `BUILD_SHARED_LIBS:BOOL=ON` to `OFF`, type `cmake .. -G "Visual Studio 14 2015 Win64"` again (don't know if this is how you would do it normally, but it seems to work).
4. Open the Visual Studio solution in the "build" folder and then build the "INSTALL" project.

I get no error, and `========== Build: 13 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========` at the end. That's it. Then when I check out the generated .lib files, they look like on my image above.

Additional information:
- I'm on Windows 7.
- Both CMake (as stated above) and Visual Studio are updated to their newest release version.
- I'm using GLFW, which CMake seems to find and identify correctly. GLFW is also in its newest release version, and also statically built.
- I never had this problem before with any statically built library, but as I said, I'm very new to this, so that might be pure chance.

What could be the reason(s) for this? Is it somehow linking all the examples, tests, tools and everything else into those .lib files, too? How can I prevent that?
Thanks in advance for any help! I'm _really_ looking forward to using this library.

**Comment by scheibel** [comment]
Thanks for reporting.

This enormous library size seems indeed strange.
On Ubuntu, I get about 17.6MB for a static release library. The dynamic release library is 6.4MB big.

Anyone on Windows who can confirm similar file sizes?

**Comment by akatray** [comment]
85mb on windows compiled with vc2013.

**Comment by cgcostume** [comment]
@Blacklands thanks for reporting, you have done everything correct: the static libs of glbinding are indeed very big. This is mainly due to the meta model which contains massive maps of strings and enums that are not optimized when compiling a static lib. Furthermore, glbinding uses templates and variadic templates which also might result in much static code. However, when linking and compiling a program with this static lib, optimization are applied and your program is similarly big as the shared library. The size of the static library is nothing that the user of your program should notice.

If this remains a relevant issue for you, look out for the next major release (3.0 is planned to be released in the next weeks), especially issue #154. Meta, logging, and more will be extracted to additional (optional) libraries which will massively reduce compile time and binary/lib size of the glbinding core.

**Comment by cgcostume** [comment]
Small additional note: linking glbinding static requires the link time optimization to run for each executable every time you change and compile it. Since this optimization takes very long time, especially on msvc compilers i prefer and suggest, at least during development, to use the dynamic library and thereby massively reduce link+compile time. Then for deployment to users use the static configuration.

Fun fact: we had about 30 min link time optimization in msvc in earlier versions of glbinding :smile: and are now down to ~3 minutes...

**Comment by mrzzzrm** [comment]
For clang3.4-1ubuntu3, static libraries, it is:
Debug: 124MB
Release: 21.4MB
MinSizeRel: 20.3MB

**Comment by Blacklands** [comment]
Hi everyone,
Thanks for the support! I had been very sure that I just did something wrong, and now I'm kind of glad that it is actually a "real" issue, not just my incompetence, lol.
What I did was exactly what you suggested; I'm using the dynamic version now, which is working just fine. :)
When that new milestone comes around, I will check the size and build speed of/with the static version, but I might possibly just stay with using the dynamic one until getting close to the end of development.
Again, much obliged. Keep up the great work!
